doc/advanced-testing-mini-course/src/Lec05/EventQueue.hs

```diff
@@ -49,19 +51,23 @@ fakeEventQueue a0 clock = do
     dequeue :: IORef Agenda -> DequeueTimeout -> IO Event
     dequeue agenda dequeueTimeout = do
       a <- readIORef agenda
-      case pop a of
-        Nothing -> case dequeueTimeout of
-          NoTimeout -> return (CommandEventE Exit)
-          Timeout _micros retry -> retry
-        Just ((time, event), a') -> case dequeueTimeout of
-          NoTimeout -> do
-            writeIORef agenda a'
-            return event
-          Timeout micros retry -> do
-            now <- cGetCurrentTime clock
-            let later = addTimeMicros micros now
-            if time <= later
-            then do
-              writeIORef agenda a'
-              return event
-            else retry
+      now <- cGetCurrentTime clock
+      let ag = case pop a of
+            Nothing -> []
+            Just ((time, event), a') ->
+              [(time, writeIORef agenda a' >> return event)]
+      let tim = case dequeueTimeout of
+            NoTimeout -> []
+            Timeout micros retry ->
+              [(addTimeMicros micros now, retry)]
+      ncr <- cgNextClientRequest clientGenerator
+      let cg = case ncr of
+            CurrentlyNoRequests -> []
+            Now e -> [(now, return e)]
+            Later t e -> [(t, return e)]
+      let allEvents = sortOn fst $ ag ++ tim ++ cg
```

Comment:
I wonder if this can be simplified, it feels like we are essentially merging three heaps here? (Same goes for `go` in the event loop.)

Reply:
Hmm, kinda but not sure exactly how to do some simplification with that. I changed it to be using an agenda that gets pushed to. Not sure it made it simpler though. I will merge for now and then we fix it later if we have something nicer.